### ISSUE TITLE
Replace scripts

### ISSUE THREAD

**Issue by ahmed-sigmalux** [o10n-x/wordpress-javascript-optimization#16]
We replace many bundled scripts with CDNJS versions using a custom plugin like:

```
foreach( $cdnjs_scripts as $script ) {
    wp_deregister_script( $script->handle );
    wp_register_script( $script->handle, $script->src, $script->deps, $script->ver, $script->footer );
    if( wp_script_is( $script->handle, 'enqueued' ) ) {
      wp_dequeue_script( $script->handle );
      wp_enqueue_script( $script->handle );
    }
  }

add_action( 'wp_enqueue_scripts', 'use_cdnjs_scripts', 10 );
```

Enabling JavaScript minification breaks our custom plugin as it appears to apply the minification and rename the scripts before our plugin fires at wp_enqueue_scripts.

At what point is this plugin replacing scripts? I imagine if we change our custom plugin to fire before that event then we can avoid the JavaScript minification from touching the CDNJS versions.

On a related note, it would be nice if you built a module that allowed replacing bundled scripts with CDN versions by WordPress script handle instead of script tag URLs.

**Comment by optimalisatie** [comment]
Hi Ahmed,

The javascript optimization plugin does not fire before `wp_enqueue_scripts`, it operates on the basis of HTML (output buffer) which means that in this case, the plugin will fetch the javascript from the CDN URLs using the WordPress HTTP API and then add the scripts to the concatenated script.

If you wish to exclude the CDN based scripts from minification and/or concatenation you will need to manually exclude the scripts via the JSON filter configuration.

**Comment by ahmed-sigmalux** [comment]
What's your reasoning for replacing the scripts via JavaScript instead of the built-in WordPress hooks? I ask because bypassing the WordPress hooks has the potential to break many other plugins.

**Comment by optimalisatie** [comment]
The scripts are replaced by PHP (WordPress filters). Scripts are loaded via javascript only when you enable asynchronous script loading.

https://www.giftofspeed.com/asynchronous-scripts/

The javascript optimization plugin loads scripts asynchronously via [little-loader](https://github.com/walmartlabs/little-loader) by Walmart Labs.

**little-loader: The Only Correct Script Loader Ever Made**
https://formidable.com/blog/2016/01/07/the-only-correct-script-loader-ever-made/

When you do not use async script loading, scripts will simply be replaced with the optimized version 'in place'. The position of the element in the HTML will remain the same, just the `src` URL will be changed to the optimized version. 

When you use concatenation without async loading then all scripts are inserted on a new position in the `<head>` of the document.

**Comment by ahmed-sigmalux** [comment]
Sorry for not explaining myself clearly. Let me try again:

I am replacing some WordPress core scripts like jquery and jquery-migrate with a custom plugin that dequeues and deregisters the local versions and then registering and queueing CDNJS versions. I would like to avoid these scripts being touched by the plugin by providing an array of registered script names, e.g., [ 'jquery', 'jquery-migrate', ]. This doesn't appear possible, however, since the plugin modifies the PHP output buffer instead of using WordPress hooks to replace scripts.

You are advising me to use the plugin JSON and provide the filenames to be excluded from processing. The problem with this method is that the script filenames can change on a regular basis, e.g., going from https://cdnjs.cloudflare.com/some/script-1.0.min.js to https://cdn.jquery.com/some/script-written-differently_1.0.minimized.js. Therefore any time a script is upgraded or replaced, I might have to go back and rewrite the plugin JSON.

Using the WordPress script names would be more developer friendly and less prone to errors because I could tell the plugin to exclude 'jquery' and 'jquery-migrate' and then only worry about changing the filenames a single time, when I replace them with a custom plugin. This could also apply to any other WordPress plugin that is upgraded and where the script filename changes, e.g., WooCommerce goes from select2.min.js to select-two.min.js.

I hope this explanation makes more sense.
